owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt

```diff
@@ -65,66 +66,84 @@ class DocumentsStorageProvider : DocumentsProvider() {
     private var requestedFolderIdForSync: Long = -1
     private var syncRequired = true
     private var currentStorageManager: FileDataStorageManager? = null
+    private lateinit var fileToUpload: OCFile
 
     override fun openDocument(documentId: String, mode: String, signal: CancellationSignal?): ParcelFileDescriptor? {
         Log_OC.d(TAG, "Trying to open $documentId in mode $mode")
-        val docId = documentId.toLong()
-        updateCurrentStorageManagerIfNeeded(docId)
 
-        var file = currentStorageManager?.getFileById(docId)
-            ?: throw FileNotFoundException("Failed to open document with id $documentId and mode $mode")
+        // If documentId == NONEXISTENT_DOCUMENT_ID only Upload is needed because file does not exist in our database yet.
+        var ocFile: OCFile
+        val uploadOnly: Boolean = documentId == NONEXISTENT_DOCUMENT_ID
 
-        if (!file.isDown) {
-            val i = Intent(context, FileDownloader::class.java).apply {
-                putExtra(FileDownloader.KEY_ACCOUNT, currentStorageManager!!.account)
-                putExtra(FileDownloader.KEY_FILE, file)
-            }
+        var accessMode: Int = ParcelFileDescriptor.parseMode(mode)
+        val isWrite: Boolean = mode.contains("w")
+
+        if (!uploadOnly) {
+            val docId = documentId.toLong()
+            updateCurrentStorageManagerIfNeeded(docId)
 
-            context?.startService(i)
+            ocFile = currentStorageManager?.getFileById(docId)
+                ?: throw FileNotFoundException("Failed to open document with id $documentId and mode $mode")
 
-            do {
-                if (!waitOrGetCancelled(signal)) {
-                    return null
+            if (!ocFile.isDown) {
+                val intent = Intent(context, FileDownloader::class.java).apply {
+                    putExtra(FileDownloader.KEY_ACCOUNT, currentStorageManager!!.account)
+                    putExtra(FileDownloader.KEY_FILE, ocFile)
                 }
-                file = currentStorageManager?.getFileById(docId)
-                    ?: throw FileNotFoundException("Failed to open document with id $documentId and mode $mode")
 
-            } while (!file.isDown)
+                context?.startService(intent)
+
+                do {
+                    if (!waitOrGetCancelled(signal)) {
+                        return null
+                    }
+                    ocFile = currentStorageManager?.getFileById(docId)
+                        ?: throw FileNotFoundException("Failed to open document with id $documentId and mode $mode")
+
+                } while (!ocFile.isDown)
+            }
+        } else {
+            ocFile = fileToUpload
```

Comment:
If we are able to know when the provider uses this method to upload or download a file, maybe we do not need to use SynchronizeFileOperation to perform a simple upload and a call to `uploadNewFile` of `TransferRequester.java` from here is enough. This way we reduce some logic in SynchronizeFileOperation by removing the `uploadOnly` parameter and do not overload it with extra responsibilities. What do you think?

You could create a transferRequest in this else condition

```
val transferRequester = TransferRequester()
transferRequester.uploadNewFile(context, currentStorageManager?.account, ..., ...)
```

Reply:
Yes, that would be definitively better

Reply:
Great, ping me when is ready so I can continue with the review, thanks!

Reply:
But we'll need to create that transfer requester later, as it is done at that commit 63e19bf97d6c11af1705597ecc7bc36b0e1a8226